### ISSUE TITLE
torch.compile silently produces wrong `nn.LayerNorm` results on MPS

### ISSUE THREAD

**Issue by jstray** [pytorch/pytorch#191976]
### 🐛 Describe the bug

`torch.compile` returns numerically incorrect output for `nn.LayerNorm` on the MPS backend. Eager matches a float64 CPU reference to ~2e-7; the compiled version diverges by up to **1.8 absolute** on inputs whose own range is roughly [-4, 4] — the error exceeds the output range.

There is no warning, no graph break, and no device fallback. The compiled module simply returns wrong numbers.

```python
import torch, torch.nn as nn

torch.manual_seed(0)
dim = 512
x = torch.randn(4, dim)
layer = nn.LayerNorm(dim)

reference = nn.LayerNorm(dim).double()
reference.load_state_dict({k: v.double() for k, v in layer.state_dict().items()})
truth = reference(x.double())                       # float64 CPU ground truth

on_mps = layer.to("mps").eval()
with torch.inference_mode():
    eager = on_mps(x.to("mps")).cpu().double()
    compiled = torch.compile(on_mps)(x.to("mps")).cpu().double()

print("eager    vs truth:", (eager - truth).abs().max().item())    # 3.3e-07
print("compiled vs truth:", (compiled - truth).abs().max().item()) # 1.8e+00
```

A self-contained script covering all dimensions and seeds is attached
(`bug_repro_mps_layernorm.py`).

## Expected behaviour

Both should agree with the float64 reference to float32 precision, ~1e-7.

## Actual behaviour

`max |output - float64_reference|`, three seeds per row:

| normalized_shape | eager | torch.compile |
|---|---|---|
| 32 | 1.2e-07, 2.4e-07, 2.4e-07 | 1.2e-07, 2.4e-07, 2.4e-07 |
| 64 | ~2e-07 | **1.7e-02**, 2.4e-07, **2.0e-02** |
| 96 | ~2e-07 | **6.7e-03, 1.7e-02, 1.1e-02** |
| 128 | ~2e-07 | **3.1e-04, 4.1e-02, 9.5e-03** |
| 256 | ~2e-07 | **4.7e-03, 6.7e-03, 1.0e-02** |
| 512 | ~3e-07 | **2.0e+00, 2.8e+00, 1.8e+00** |

Eager is correct throughout. Onset is **input-dependent**: `dim=64` fails for
seeds 0 and 2 but not seed 1. `dim >= 96` failed on every seed tried, `dim=32`
on none, and the error grows with the dimension.

## Scope — only LayerNorm

Each module tested individually on MPS, eager vs compiled:

| module | max diff |
|---|---|
| `nn.Linear` | 0.0 |
| `nn.GELU` | 0.0 |
| `nn.Softmax(dim=-1)` | 0.0 |
| `nn.MultiheadAttention`, no mask | 0.0 |
| `nn.MultiheadAttention`, `key_padding_mask` | 0.0 |
| **`nn.LayerNorm`** | **0.174** |
| `nn.TransformerEncoderLayer` | 0.341 — contains LayerNorm |

Other observations:

- **fp32 and fp16 are affected equally**, so this is not a precision tradeoff and is distinct from #66707 (layer_norm needing fp32 accumulation for fp16).
- Reproduces with `elementwise_affine=True` and `False`.
- `torch._dynamo.utils.counters["graph_break"]` is empty — it compiles cleanly.
- Both outputs are on `mps:0`, so it is not a silent CPU fallback.
- Eager being correct rules out the MPS "fast math" behaviour in #84936, which would affect both paths.

## Possible mechanism

The `torch.compile` on MPS tracker (#150121) notes that **multi-stage Welford reductions are not implemented**. LayerNorm computes mean and variance by a Welford reduction over `normalized_shape`, and the failure appearing between 32 and 96 elements — and worsening with size — is consistent with a reduction that is correct while it fits one stage and wrong once it needs more. #152155 reports a related multistage-reduction failure, though that one raises a shader error rather than returning wrong values.

## Related issues

- #150121 — torch.compile on MPS progress tracker (open). This probably belongs under it.
- #143182 — same symptom class, `nn.LayerNorm` silently wrong under Inductor,  but **CPU only** and **closed**, with a much smaller error (7e-4 vs 1.8).
- #150629, #150454 — RMSNorm on MPS under torch.compile: outputs zeros, and  generates invalid shaders. Adjacent normalisation-layer failures.
- #152155 — multistage reductions on MPS generating out-of-scope variables.

## Real-world impact

Found while embedding `sentence-transformers/all-MiniLM-L6-v2` (6-layer BERT, hidden size 384). The compiled encoder differed from eager by **11.1 absolute** on `last_hidden_state` values spanning [-5.2, 5.3], corrupting every embedding with nothing to indicate a problem — no warning, no error, plausible-looking output.

On that model `torch.compile` was also **~7x slower** than eager (416 vs 2,885 texts/s at batch 512, fp16), so there is no speed incentive to offset the correctness risk.

The tracker does describe MPS support as an early prototype, so a failure is not surprising. Filing because this one is *silent*: a user who benchmarks `torch.compile`, sees it is slower, and moves on is fine — but a user who sees a speedup on a different model and adopts it has no signal that the numbers are wrong.

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20 @chauhang @penguinwu (MPS), @jansel @desertfire (torch.compile)

[bug_repro_mps_layernorm.py](https://github.com/user-attachments/files/30677340/bug_repro_mps_layernorm.py)

### Versions

PyTorch version: 2.7.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5.1 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.0 (main, Feb  4 2025, 11:41:54) [Clang 16.0.0 (clang-1600.0.26.6)] (64-bit runtime)
Python platform: macOS-26.5.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M4 Pro

Versions of relevant libraries:
[pip3] numpy==2.2.2
[pip3] torch==2.7.1
[conda] Could not collect

**Comment by Isalia20** [comment]
Can't reproduce on main:
```
eager    vs truth: 3.9349332636717804e-07
compiled vs truth: 3.989298664741625e-07
```

Please update pytorch to latest version or the nightly one and check again

**Comment by malfet** [comment]
> torch==2.7.1

Please update to a more recent version of torch (2.13 works fine)

**Comment by jstray** [comment]
Confirmed fixed in 2.13.0. Thanks, and apologies for the noise.
